invenio_communities/requests/user_moderation/actions.py

```diff
@@ -1,26 +1,114 @@
 # -*- coding: utf-8 -*-
 #
 # Copyright (C) 2023 CERN.
+# Copyright (C) 2023 TU Wien.
 #
 # Invenio-communities is free software; you can redistribute it and/or modify
 # it under the terms of the MIT License; see LICENSE file for more details.
 """Communities user moderation actions."""
 
+from collections import defaultdict
+
 from invenio_access.permissions import Identity, system_identity
-from invenio_records_resources.services.uow import RecordIndexOp, unit_of_work
+from invenio_pidstore.errors import PIDDoesNotExistError
 from invenio_search.engine import dsl
+from invenio_vocabularies.proxies import current_service as vocab_service
 
 from invenio_communities.proxies import current_communities
 
+from ...communities.records.systemfields.deletion_status import (
+    CommunityDeletionStatusEnum,
+)
+
+
+def _get_communities_for_user(user_id):
+    """Helper function for getting all communities with the user as the sole owner.
+
+    Note: This function performs DB queries yielding all communities with the given
+    user as the sole owner (which is not hard-limited in the system) and performs
+    service calls on each of them. Thus, this function has the potential of being a very
+    heavy operation, and should not be called as part of the handling of an
+    HTTP request!
+    """
+    comm_cls = current_communities.service.record_cls
+    comm_model_cls = comm_cls.model_cls
+    mem_cls = current_communities.service.members.record_cls
+    mem_model_cls = mem_cls.model_cls
+
+    # collect the owners for each community
+    comm_owners = defaultdict(list)
+    for comm_owner in [
+        mem_cls(m.data, model=m)
+        for m in mem_model_cls.query.filter(mem_model_cls.role == "owner").all()
+    ]:
+        comm_owners[comm_owner.community_id].append(comm_owner)
+
+    # filter for communities that are owned solely by the user in question
+    relevant_comm_ids = [
+        comm_id
+        for comm_id, owners in comm_owners.items()
+        if len(owners) == 1 and str(owners[0].user_id) == user_id
+    ]
```

Comment:
i have a feeling that this should be doable with a `group_by` and/or `having` for better performance

Reply:
I think we have a helper function to [get all memberships of a user](https://github.com/inveniosoftware/invenio-communities/blob/cfc230e8fb55f9fba8b75b45b92ba01705752f35/invenio_communities/members/records/api.py#L105). It returns though all memberships and not filtered by a specific role but it could be improved. Ping @ntarocco I think in this case the performance issue with getting all communities of a user from DB is done on demand so it should not affect overall performance right?

Reply:
to be discussed, it seems to me that it is not a blocker for now